src/ucx_plugin.c

```diff
@@ -519,24 +519,196 @@ ncclResult_t nccl_ucx_deregmr(void* comm, void* mhandle) {
   return ncclSuccess;
 }
 
+ncclResult_t ucx_send_check(ucx_send_comm_t *comm) {
+  ucp_tag_message_h     msg_tag;
+  ucp_tag_recv_info_t   info_tag;
+  ucx_request_t         *req;
+  connect_msg_t         *msg;
+  ucp_ep_params_t       ep_params;
+
+  ucp_worker_progress(comm->worker);
+
+  msg_tag = ucp_tag_probe_nb(comm->worker, comm->ctag, tag_mask, 1, &info_tag);
+  if (msg_tag == NULL) {
+    return ncclSuccess;
+  }
+
+  msg = malloc(info_tag.length);
+  req = ucp_tag_msg_recv_nb(comm->worker, msg, info_tag.length, ucp_dt_make_contig(1), msg_tag, recv_handler);
+
+  if (UCS_PTR_IS_ERR(req)) {
+    WARN("Unable to receive connect msg (%s)", ucs_status_string(UCS_PTR_STATUS(req)));
+  }
+  else {
+    while (req->completed == 0) {
+      ucp_worker_progress(comm->worker);
+    }
+    req->completed = 0;
+    ucp_request_release(req);
+  }
+
+  ep_params.field_mask = UCP_EP_PARAM_FIELD_REMOTE_ADDRESS;
+  ep_params.address    = (ucp_address_t*)(msg + 1);
+  UCXCHECK(ucp_ep_create(comm->worker, &ep_params, &comm->ep));
+  comm->ready = 1;
+  free(msg);
+
+  return ncclSuccess;
+}
+
+ncclResult_t ucx_recv_check(ucx_recv_comm_t *comm) {
+  ucp_address_t *my_addr;
+  size_t        local_addr_len;
+  size_t        msg_len;
+
+  if (comm->connect_req == NULL) {
+    NCCLCHECK(ucx_worker_get_netaddress(comm->worker, &my_addr, &local_addr_len));
+    nccl_ucx_add_ep(comm->worker, comm->fd);
+    msg_len = sizeof(connect_msg_t) + local_addr_len;
+    comm->msg = calloc(1, msg_len);
+    comm->msg->addr_len = local_addr_len;
+    memcpy(comm->msg + 1, my_addr, local_addr_len);
+    comm->connect_req = ucp_tag_send_nb(comm->ep, comm->msg, msg_len, ucp_dt_make_contig(1), comm->ctag, send_handler);
+
+    if (UCS_PTR_IS_ERR(comm->connect_req)) {
+      WARN("Unable to send connect message");
+      return ncclSystemError;
+    } else if (comm->connect_req == NULL){
+      comm->ready = 1;
+      free(comm->msg);
+    }
+
+    free(my_addr);
+  }
+  else{
+    if (comm->connect_req->completed == 1) {
+      comm->ready = 1;
+      comm->connect_req->completed = 0;
+      ucp_request_release(comm->connect_req);
+      free(comm->msg);
+    }
+    else {
+      ucp_worker_progress(comm->worker);
+    }
+  }
+  return ncclSuccess;
+}
+
+
 ncclResult_t nccl_ucx_isend(void *send_comm, void *data, int size, void *mhandle, void **request) {
-  WARN("NET/UCX: not implemented");
-  return ncclInternalError;
+  ucx_send_comm_t   *comm = (ucx_send_comm_t *)send_comm;
+  volatile uint32_t *head = &comm->fifo_head;
+  volatile uint32_t *tail = comm->fifo_tail;
+  ucx_request_t     *req;
+
+  if (comm->ready == 0) {
+    ucx_send_check(comm);
+  }
+
+  if (comm->ready == 0) {
```

Comment:
instead of the same condition here, can we return appropriate return value from ucx_send_check and check this at line:605